include/simdjson/generic/ondemand/array-inl.h

```diff
@@ -72,6 +72,24 @@ simdjson_really_inline simdjson_result<array_iterator> array::end() noexcept {
   return array_iterator(iter);
 }
 
+simdjson_really_inline simdjson_result<size_t> array::count_elements() & noexcept {
+  // The count_elements method should always be called before you have begun
+  // iterating through the array.
+  const bool at_start{iter.is_at_container_start()};
+  SIMDJSON_ASSUME(at_start); // we do not want SIMDJSON_ASSUME to have side-effects.
```

Comment:
I've been trying not to use SIMDJSON_ASSUME to validate API misuse, only to validate internal assumptions. If the API cannot be called unless it's at the start, we should return an error code; alternatively, we could make it restart the array or change the contract to count *remaining* elements. I think I'd prefer the error, because then it would fail if you tried to use it *while* iterating the array, which seems like a common mistake to make that would either be unperformant or cause weird iterator invalidation.

I expect an error check to be free (compiled away) in most cases anyway, as long as `is_at_container_start()` can be determined at compile time--which it should be if you have just opened the array.

Reply:
To be fixed.

Reply:
Done. Replacing with...

```C++
if(!iter.is_at_container_start()) { return OUT_OF_ORDER_ITERATION; }
```

Reply:
guarded with a macro check.